add type for mxc://

jsparber opened the pull request:
Fixes: #26

jplatte commented on ruma-identifiers-validation/src/mxc.rs:
```diff
@@ -0,0 +1,24 @@
+const PROTOCOL: &str = "mxc://";
+
+pub fn validate(uri: &str) -> Result<(&str, &str), Error> {
+    if let Some(prefix_index) = uri.find(PROTOCOL) {
```
I'd prefer if this used early returns for the error cases to reduce indentation. Also I just increased the MSRV to 1.45 so you can use `strip_prefix` for this.

jsparber replied:
Oh no, i removed it, since it wasn't available in CI. I will add it back.

jsparber replied:
done.

jplatte commented on ruma-identifiers/src/mxc.rs:
```diff
@@ -0,0 +1,119 @@
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Mxc {
+    server_name: ServerNameBox,
+    media_id: MediaIdBox,
```
Since string conversions are much more common than accessing individual parts of identifiers we've generally been storing the entire string as one thing. Also I'm planning to soon remove all additional metadata entirely such that the ID types are DSTs (dynamically-sized types), but you don't need to do the same in this PR.

jsparber replied:
I think in the case of the `mxc` we receive an url, but use pretty much always the parts. Have a look at the media requests, they all require the server name and the media id as separate parameter.

jplatte replied:
Hrm, this would be the only type that can't become a ZST then. I'll have to think about it...

jsparber replied:
I guess modifying the media requests to accept the `mxc://`  directly isn't really a viable option, right?

jplatte replied:
It might be viable, but I think that's tangential to how we store the `mxc` URLs.

jsparber replied:
It would remove the need of accessing the parts of the `mxc://`.

jplatte commented on ruma-identifiers/src/opaque_ids.rs:
```diff
@@ -28,6 +28,14 @@ opaque_identifier! {
+    ///
+    /// Media identifiers in Matrix are opaque character sequences of `[a-zA-Z_-]`. This type is provided
+    /// simply for its semantic value.
+    pub type MediaId;
```
I'm not sure we need this extra type. But if we want to keep it, we should probably also enforce the character range.

jsparber replied:
I added it since media requests expect the media id separate from the server name.

jsparber replied:
i removed it.

jsparber committed:
add type for mxc://

jsparber committed:
ruma-client-api: Use Mxc type

This also fixes the wrong mxc url in the test for joined_members. Mxc
urls don't use a ":" as seperator between server name and media id.

jplatte commented on ruma-identifiers/src/mxc.rs:
```diff
@@ -0,0 +1,119 @@
+
+    /// Returns the server name of this mxc://.
+    pub fn server_name(&self) -> &ServerName {
+        self.server_name.as_ref()
```
This can also just be `&self.server_name`.

jplatte commented on ruma-identifiers/src/mxc.rs:
```diff
@@ -0,0 +1,119 @@
+
+impl fmt::Debug for Mxc {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&self.to_string())
```
I don't like that this allocates a temporary string when `Display` doesn't. Can you add an inherent `_fmt` that does what `Display::fmt` does now and is called by both `Debug` and `Display`?

jplatte commented on ruma-client-api/src/r0/membership/joined_members.rs:
```diff
@@ -80,13 +80,13 @@ mod test {
                 avatar_url: Some(avatar_url),
             } if display_name == "alice"
-                && avatar_url == "mxc://localhost:wefuiwegh8742w"
+                && &avatar_url.to_string() == "mxc://localhost/wefuiwegh8742w"
```
Is the extra `&` needed here?

jsparber replied:
We can also use `String::from()` for the left side of the comparison.

jplatte replied:
I would have expected `(foo: Sting) == "string constant"` to compile. If it doesn't, let's keep the code as-is.

jsparber replied:
actually it does compile without it.

jplatte commented:
Thanks!
